## <b><i>Estructurando un Proyecto de DS (Parte III)</i></b>
Autor: <i>Emanuel Benitez</i><br>
Fecha: <i>28-10-2023</i>

#### <b>Consignas</b>

[Complementar el trabajo realizado]

<ol>
<li>Elegir un método de feature selection para reducir la dimensionalidad del conjunto de datos.</li>
<li>Elegir un algoritmo de regresión o clasificación para entrenar con los datos elegidos.</li>
<li>Hacer un cálculo de metricas para validar el modelo.</li>
<li>Generar conclusiones en base a los resultados obtenidos.</li>  
</ol>

<b><u>Contenido</u></b>
<ol>
    <li>Introducción</li>
        - Abstracto, descripción del conjunto de datos y preguntas a resolver
    <li> Análisis Exploratorio de Datos </li>
        - Exploración detallada que incluye: estadísticas descriptivas, visualizaciones y patrones clave observados.
    <li> <b>Selección de características (Feature Selection)</b> </li>
        - Descripción del metodo de selección de características utilizado, caracteristicas elegidas y justificación.
    <li> <b>Modelado</b> </li>
        - Ajuste del modelo y configuración a partir de las caracteristicas elegidas.
    <li> <b>Validación del modelo</b></li>
        - Evaluación del rendimiento del modelo e interpretación de metricas
    <li> <b>Resultados y conclusiones a partir del modelo</b> </li>
        - Resultados y conclusiones basados en el rendimiento del modelo
    <li> Respuestas a las preguntas planteadas </li>
        - Respuestas en base al análisis exploratorio y el modelado
    <li> Recomendaciones basadas en los descubrimientos (insights) </li>
    <li> Conclusión </li>
    
</ol>

# <center>Introducción</center>
<center>Abstracto, descripción del conjunto de datos y preguntas a resolver</center>
<br>
<center>
  <img src="https://raw.githubusercontent.com/emanuelbe1/ciencia_de_datos_repasos/707fa795eed2a1805e2c8c1a7d9cc55a77fa033e/Coderhouse/ProyectoDS_ParteII/2023-10-17_11.58.09_upscaled.jpg" 
       alt="Collage de distintos tipos de instrumentos con una guitarra en el centro. Imagen generada por herramienta de inteligencia artificial." 
       width="600" >
</center>


### <center><b>Abstracto</b></center>
<p>&nbsp;La música es un elemento central en la vida de las personas y una poderosa fuerza en la cultura y el entretenimiento. El surgimiento de plataformas de streaming como Spotify y la ubicuidad de YouTube como una plataforma de videos han transformado la forma en que experimentamos y consumimos música. En este contexto, el análisis de datos se convierte en una herramienta fundamental para comprender las tendencias y patrones que influyen en la industria musical, desde la popularidad de una canción hasta las preferencias del público.</p>
<br>
<br>

<b>Audiencia</b>
<p>&nbsp;La audiencia de este estudio abarca tanto a los profesionales de la industria musical, que pueden beneficiarse de una comprensión más profunda de las métricas que impulsan el éxito de una canción, como a la comunidad de investigadores y entusiastas de la música, interesados en desentrañar las relaciones entre las características musicales y su recepción por parte del público. Además, este análisis puede ser valioso para aquellos que buscan perfeccionar estrategias de promoción y toma de decisiones relacionadas con la música en la era digital.</p>
<br>
<br>

<b>Motivación</b>
<p>&nbsp;En resumen, este estudio se motiva por la necesidad de arrojar luz sobre los factores que influyen en el desempeño de una canción en las plataformas de streaming y su recepción en YouTube. Su audiencia abarca desde profesionales de la industria hasta apasionados de la música y tiene como objetivo proporcionar información valiosa y perspicaz en el emocionante y siempre cambiante mundo de la música digital.</p>

<br>
<br>

<b>Preguntas a resolver</b><br>

   1. <i>¿Hay una relación entre el número de reproducciones en Spotfy y la adecuación para el baile de la canción?</i><br>
    <p>&nbsp;En esta cuestión, consideramos el número de reproducciones como indicativo de la popularidad de una canción. Se podría reformular la pregunta a: ¿tienden las canciones "bailables" a ser más populares?</p>
    <br><br>
   2. <i>¿Existen variables altamente correlacionadas con el número de comentarios en Youtube?</i><br>
    <p>&nbsp;En esta pregunta se busca encontrar los factores correlacionados con un indicativo de fama (o infamia) de una canción: se toma el número de comentarios en Youtube cómo un indicativo de popularidad.</p>
    <br><br>
   3. <i>¿Qué tan populares son, en terminos de reproducciones de Spotify, las canciones instrumentales?</i><br>
   <p>&nbsp;En esta cuestión se explora si hay alguna relación ¿tienden a ser menos populares las canciones instrumentales? ¿tienden a ser más escuchadas?</p>
    <br><br>


<br>
<br>

<details><summary><b>Hipótesis (a completar)</b></summary>

- Hipótesis 1: Las canciones tristes ('valence' cercano a 0) tienden a tener menos uso de la voz (mayor 'instrumentalness').
- Hipótesis 2: Las canciones "positivas" o "alegres" ('valence' cercano a 1) tienden a una mayor popularidad en terminos de reproducciones de Spotify/Youtube.
</details>
<br>

<center><img src="https://raw.githubusercontent.com/emanuelbe1/ciencia_de_datos_repasos/main/Coderhouse/ProyectoDS_ParteII/img/if_spotify_then_youtube.png" alt="Logos de Spotify y Youtube, uno al lado de otro" width=600></center>

#### <center><b><u>Sobre el conjunto de datos</u>:</b> <i>'Spotify and Youtube'</i></center>


<details><summary><b><u>Descripción</u></b></summary><br>
<font size = "3">&nbsp;Este conjunto de datos contiene estadisticas de canciones para sus versiones lanzadas en Spotify. También contiene el número de visitas en su video oficial, alojado en la plataforma Youtube. En el contenido se incluyen <b>26 variables</b> para cada uno de las canciones recolectadas de Spotify. Entre estás variables se incluyen: nombre de la canción (Track), nombre del artista (Artist), album donde se contiene la canción (Album), adecuación para el baile (Danceability), tonalidad de la canción (Key), tiempo estimado de la canción (Tempo), duración de la canción en milisegundos (Duration_ms), titulo del video en Youtube (Title), entre otras.</font>

[Fuente de los datos](https://www.kaggle.com/datasets/salvatorerastelli/spotify-and-youtube)</details>
<details><summary><b><u>Índice de variables</u></b></summary><br>
<font size = "3">&nbsp;La siguiente tabla contiene una descripción exhaustiva para cada una de las columnas en los datos.</font>

<font size = "2"><b>Track</b>: Nombre de la canción en Spotify<br>
<b>Artist</b>: Nombre del artista<br>
<b>Url_spotify</b>: URL del artista en Spotify<br>
<b>Álbum</b>: Álbum en el que se encuentra la canción en Spotify<br>
<b>Album_type</b>: Indica si la canción se ha lanzado en Spotify como sencillo o en un álbum<br>
<b>Uri</b>: Un enlace de Spotify utilizado para encontrar la canción a través de la API<br>
<b>Danceability</b>: Describe cuán adecuada es una pista para el baile basándose en una combinación de elementos musicales, incluyendo tempo, estabilidad del ritmo, fuerza del ritmo y regularidad general. Un valor de 0.0 es menos bailable y 1.0 es muy bailable<br>
<b>Energy</b>: Es una medida de 0.0 a 1.0 que representa una medida perceptual de intensidad y actividad. Típicamente, las pistas energéticas se sienten rápidas, ruidosas y enérgicas<br>
<b>Key</b>: La tonalidad en la que se encuentra la canción. Los enteros se mapean a notas utilizando notación estándar de clases de tonos. Por ejemplo, 0 = Do, 1 = Do♯/Re♭, 2 = Re, y así sucesivamente. Si no se detectó una tonalidad, el valor es -1<br>
<b>Loudness</b>: El volumen general de una pista en decibelios (dB). Los valores de volumen se promedian en toda la pista y son útiles para comparar el volumen relativo de las pistas<br>
<b>Speechiness</b>: Detecta la presencia de palabras habladas en una pista. Cuanto más parecido a palabras habladas sea el registro (por ejemplo, un programa de entrevistas, audiolibro, poesía), más cercano a 1.0 será el valor del atributo<br>
<b>Acousticness</b>: Una medida de confianza de 0.0 a 1.0 de si la pista es acústica. 1.0 representa una alta confianza de que la pista es acústica<br>
<b>Instrumentalness</b>: Predice si una pista no contiene vocales. Los sonidos "Ooh" y "aah" se consideran instrumentales en este contexto. Las pistas de rap o palabras habladas son claramente "vocales". Cuanto más cercano sea el valor de instrumentalidad a 1.0, mayor será la probabilidad de que la pista no contenga contenido vocal<br>
<b>Liveness</b>: Detecta la presencia de una audiencia en la grabación. Valores de en_vivo más altos representan una mayor probabilidad de que la pista se haya interpretado en vivo<br>
<b>Valence</b>: Una medida de 0.0 a 1.0 que describe la positividad musical transmitida por una pista. Pistas con alto valor de positividad suenan más positivas (por ejemplo, alegres, alegres, eufóricas), mientras que pistas con bajo valor de positividad suenan más negativas (por ejemplo, tristes, deprimidas, enojadas)<br>
<b>Tempo</b>: El tempo estimado general de una pista en pulsos por minuto (BPM)<br>
<b>Duration_ms</b>: La duración de la pista en milisegundos<br>
<b>Stream</b>: Número de transmisiones de la canción en Spotify<br>
<b>Url_youtube</b>: URL del video vinculado a la canción en YouTube, si existe<br>
<b>Title</b>: Título del videoclip en YouTube<br>
<b>Channel</b>: Nombre del canal que ha publicado el video<br>
<b>Views</b>: Número de vistas<br>
<b>Likes</b>: Número de "Me gusta"<br>
<b>Comments</b>: Número de comentarios<br>
<b>Description</b>: Descripción del video en YouTube<br>
<b>Licensed</b>: Indica si el video representa contenido con licencia, lo que significa que el contenido fue cargado a un canal vinculado a un socio de contenido de YouTube y luego reclamado por ese socio<br>
<b>official_video</b>: Valor booleano que indica si el video encontrado es el video oficial de la canción<br>
</font>
<br></details>

<hr>

# <center><b>Análisis exploratorio de datos</b></center>
<center> Estadísticas descriptivas, visualizaciones y patrones clave observados.
</center>

In [1]:
# Comenzamos importando las librerias necesarias para el análisis de datos

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from mlxtend.feature_selection import SequentialFeatureSelector as SFS
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error, r2_score


url='https://drive.google.com/file/d/1at4CyKa-PyJhTrLK-SHMJAZ1_PNnQ_gI/view?usp=drive_link'
url='https://drive.google.com/uc?id=' + url.split('/')[-2]

spotifyYoutube = pd.read_csv(url)
# Para hacer más practico el análisis y debido a que no voy a utilizar los valores en estás columnas
# decidi excluir las variables: "Url_spotify", "Uri", "Url_youtube", "Description".
columnas_a_eliminar = ["Unnamed: 0","Url_spotify", "Uri", "Url_youtube", "Description"]

spotifyYoutube = spotifyYoutube.drop(columns=columnas_a_eliminar)



n_entradas = spotifyYoutube.shape[0]
n_columnas = spotifyYoutube.shape[1]

print(f"Hay \033[1m{n_entradas} entradas\033[0m o registros, y \033[1m{n_columnas} columnas\033[0m o variables.\n")
print(f"\033[1mLista de columnas:\033[0m\n\n {spotifyYoutube.columns}")
print("___"*15)

Hay 20718 entradas o registros, y 23 columnas o variables.

Lista de columnas:

 Index(['Artist', 'Track', 'Album', 'Album_type', 'Danceability', 'Energy',
       'Key', 'Loudness', 'Speechiness', 'Acousticness', 'Instrumentalness',
       'Liveness', 'Valence', 'Tempo', 'Duration_ms', 'Title', 'Channel',
       'Views', 'Likes', 'Comments', 'Licensed', 'official_video', 'Stream'],
      dtype='object')
_____________________________________________


In [2]:
spotifyYoutube.describe()

Danceability        Energy           Key      Loudness   Speechiness  \
count  20716.000000  20716.000000  20716.000000  20716.000000  20716.000000   
mean       0.619777      0.635250      5.300348     -7.671680      0.096456   
std        0.165272      0.214147      3.576449      4.632749      0.111960   
min        0.000000      0.000020      0.000000    -46.251000      0.000000   
25%        0.518000      0.507000      2.000000     -8.858000      0.035700   
50%        0.637000      0.666000      5.000000     -6.536000      0.050500   
75%        0.740250      0.798000      8.000000     -4.931000      0.103000   
max        0.975000      1.000000     11.000000      0.920000      0.964000   

       Acousticness  Instrumentalness      Liveness       Valence  \
count  20716.000000      20716.000000  20716.000000  20716.000000   
mean       0.291535          0.055962      0.193521      0.529853   
std        0.286299          0.193262      0.168531      0.245441   
min        0.000001          0.000000      0.014500      0.000000   
25%        0.045200          0.000000      0.094100      0.339000   
50%        0.193000          0.000002      0.125000      0.537000   
75%        0.477250          0.000463      0.237000      0.726250   
max        0.996000          1.000000      1.000000      0.993000   

              Tempo   Duration_ms         Views         Likes      Comments  \
count  20716.000000  2.071600e+04  2.024800e+04  2.017700e+04  2.014900e+04   
mean     120.638340  2.247176e+05  9.393782e+07  6.633411e+05  2.751899e+04   
std       29.579018  1.247905e+05  2.746443e+08  1.789324e+06  1.932347e+05   
min        0.000000  3.098500e+04  0.000000e+00  0.000000e+00  0.000000e+00   
25%       97.002000  1.800095e+05  1.826002e+06  2.158100e+04  5.090000e+02   
50%      119.965000  2.132845e+05  1.450110e+07  1.244810e+05  3.277000e+03   
75%      139.935000  2.524430e+05  7.039975e+07  5.221480e+05  1.436000e+04   
max      243.372000  4.676058e+06  8.079649e+09  5.078865e+07  1.608314e+07   

             Stream  
count  2.014200e+04  
mean   1.359422e+08  
std    2.441321e+08  
min    6.574000e+03  
25%    1.767486e+07  
50%    4.968298e+07  
75%    1.383581e+08  
max    3.386520e+09

#### <b>Estadisticas descriptivas de variables númericas</b>

<p> Al aplicar la función <i>.describe</i> para hacer un análisis de las estadisticas descriptivas de las variables númericas, se puede observar que:</p>

<ul>
    
<li><b>Danceability</b>: El promedio es aproximadamente de 0.62, con un valor para el Q1 de 0.518, lo que sugiere que un 75% de las canciones tienen una adecuación para el baile mayor a este valor: el histograma sera asimetrico hacia la derecha y la música tiende a juzgarse más adecuada para el baile.</li>
    
<li><b>Energy</b>: Con valores similares, "Energy" (si la canción se percibe energica, rapida y ruidosa) tiende a tener un valor mayor a 0.507 (el 75% de las canciones), con una media de 0.635250. El histograma sera, también, asimetrico hacia la derecha y la música tiende a juzgarse o percibirse más energica.</li>
    
<li><b>Key</b>: Para la columna "Tonalidad", se observa que hay una variación estandar muy alta, de 3.576. Sin embargo, en este conjunto de datos se habla de que, cúando no se puede calcular la tonalidad, se coloca el valor de "-1". Antes tomar decisiones en base a esta columna, se debe hacer un procesamiento donde se excluyan los valores atipicos (-1).</li>
    
<li><b>Tempo</b>: El promedio de canciones, es de 120. Esto tiene sentido, ya que la metrica utilizada normalmente para seguir el pulso de metrónomo es una negra que normalmente se implica de 120 pulsos por minutos. Sin embargo, hay una variación de casí 30 pulsos por minuto (29.57). Esto cambia mucho la dinamica de una composición: un tempo de 150 es normalmente utilizado en música electronica rapida (trance, hardcore), mientras que un tempo de entre 80 a 90 BPM es normalmente utilizado en generos cómo Lo-Fi, Chill out y Vaporwave, que generan un ambiente sin distracciones. El pop, genero muy popular, normalmente usa un tempo entre 90-135, lo que explicaria en gran parte el promedio presente en esta columna. Sin embargo, antes de realizar sugerencias en base a los datos, se debe investigar más a fondo.</li>
    
</ul>
<br>

#### <b>Descripción de las variables categóricas</b>
<p>&nbsp;Con el siguiente código vamos a ver los primeros 5 resultados más frecuentes y el número de valores únicos para cada columna categorica de interés.</p>

In [3]:
variables_categoricas = spotifyYoutube.select_dtypes(include=['object']) #selecciono los variables categóricas

columnas_a_contar = ['Artist', 'Album_type', 'Channel', 'official_video']
for columna in columnas_a_contar:
    conteo_valores = variables_categoricas[columna].value_counts()
    n_valores_unicos = len(conteo_valores)
    
    print(conteo_valores.sort_values(ascending=False).head(5))
    print(f"\nNúmero de valores únicos en la columna '\033[1m{columna}\033[0m': {n_valores_unicos}")
    print("_____"*15, "\n")

Gorillaz                     10
London Symphony Orchestra    10
Michael Jackson              10
P!nk                         10
Eminem                       10
Name: Artist, dtype: int64

Número de valores únicos en la columna 'Artist': 2079
___________________________________________________________________________ 

album          14926
single          5004
compilation      788
Name: Album_type, dtype: int64

Número de valores únicos en la columna 'Album_type': 3
___________________________________________________________________________ 

T-Series              238
SonyMusicIndiaVEVO     95
SonyMusicSouthVEVO     83
Zee Music Company      65
RHINO                  62
Name: Channel, dtype: int64

Número de valores únicos en la columna 'Channel': 6714
___________________________________________________________________________ 

True     15723
False     4525
Name: official_video, dtype: int64

Número de valores únicos en la columna 'official_video': 2
___________________________________

### <b>Visualizaciones para variables de interés</b>

##### <b>Número de reproducciones (Spotify) y adecuación para el baile</b>
<p>Para intentar responder a la primer pregunta, cree un gŕafico de puntos con la variable de bailabilidad ('Danceability') y el número de stream's en Spotify. Luego imprimi las estadísticas descriptivas de: media, mediana, moda, y correlación.</p>
<img src = "https://raw.githubusercontent.com/emanuelbe1/ciencia_de_datos_repasos/main/Coderhouse/imagenes/danceability_vs_streams.jpg"></img>

In [4]:
danceability_mean = spotifyYoutube['Danceability'].mean()
danceability_median = spotifyYoutube['Danceability'].median()
danceability_mode = spotifyYoutube['Danceability'].mode()[0]
danceability_stream_corr = spotifyYoutube[['Danceability', 'Stream']].corr()['Stream'][0]

print(f"Media de la columna 'Danceability': {danceability_mean}")
print(f"Mediana de la columna 'Danceability': {danceability_median}")
print(f"Moda de la columna 'Danceability': {danceability_mode}") 
print(f'Correlación entre número de "Streams" y "Danceability": {danceability_stream_corr}') 
print("_____" * 15)


Media de la columna 'Danceability': 0.6197774522108509
Mediana de la columna 'Danceability': 0.637
Moda de la columna 'Danceability': 0.687
Correlación entre número de "Streams" y "Danceability": 0.07337514004570202
___________________________________________________________________________


##### <b>Variables correlacionadas a comentarios en Youtube: matriz de correlación</b>
<p>&nbsp;Para ver las distintas variables correlacionadas con el número de comentarios en Youtube, he decidido usar una visualización del tipo mapa de calor ("heatmap") para todas las variables del conjunto de datos.</p>
<img src="https://raw.githubusercontent.com/emanuelbe1/ciencia_de_datos_repasos/main/Coderhouse/imagenes/correlacion_sy.jpg"></img>

<p>&nbsp;Observando la matriz de correlación poder ver que podria existir una relación lineal entre la cantidad de vistas, me gusta (likes) y reproducciones en Spotify (stream) con la variable "Comments". Esto tiene sentido, ya que a más visitas en un video músical, mayor es la probabilidad de recibir un comentario. Estas mismas visitas se explicarian con la cantidad de reproducciones en Spotify: el usuario entra al video oficial de la canción (que aumenta las visitas) y luego comenta y da su like. Estás tres variables juntas podrían tener el potencial de describir la cantidad de comentarios que puede recibir el video de una canción</p>

<hr>

#####  <b>Gráfico de dispersión: instrumentalidad por reproducciones en Spotify</b>
 <p>&nbsp;Para responder a la tercer pregunta, he creado una visualización de las reproducciones en spotify y la instrumentalidad de la canción. Separe las canciones con menos probabilidad de tener contenido vocal (mayores a 0.5 de instrumentalness) en color azul, y con color rojo las que más probabilidad tienen de contener vocalizaciones y no ser instrumentales (menos de 0.5 de instrumentalness). </p>
 
 <img src="https://raw.githubusercontent.com/emanuelbe1/ciencia_de_datos_repasos/main/Coderhouse/imagenes/streamsvsinstrumental.jpg"></img>

In [5]:
instrumentalness_mean = spotifyYoutube['Instrumentalness'].mean()
instrumentalness_median = spotifyYoutube['Instrumentalness'].median()
instrumentalness_mode = spotifyYoutube['Instrumentalness'].mode()[0]
instrumentalness_stream_corr = spotifyYoutube[['Instrumentalness', 'Stream']].corr()['Stream'][0]


print(f"Media de la columna 'Instrumentalness': {instrumentalness_mean}")
print(f"Mediana de la columna 'Instrumentalness': {instrumentalness_median}")
print(f"Moda de la columna 'Instrumentalness': {instrumentalness_mode}") 
print(f'Correlación entre número de "Streams" e "Instrumentalness": {instrumentalness_stream_corr}') 
print("_____" * 15) 


Media de la columna 'Instrumentalness': 0.0559615577143274
Mediana de la columna 'Instrumentalness': 2.405e-06
Moda de la columna 'Instrumentalness': 0.0
Correlación entre número de "Streams" e "Instrumentalness": -0.0919418357027163
___________________________________________________________________________


## <center>Selección de características</center>

<center>Descripción del metodo de selección de características utilizado, caracteristicas elegidas y justificación.</center>

<p> Voy a usar buscar las variables que puedan de alguna manera, predecir la popularidad de una canción. Indicadores de popularidad podran ser:</p>

- La columna 'Stream', para las reproducciones en Spotify

- La columna 'Views', para las visualizaciones en Youtube
    
<p>En mi conjunto de datos, las variables cuantitativas que describen una melodia son:</p>

- Danceability

- Energy

- Loudness

- Speechiness

- Acousticness

- Instrumentalness

- Valence

- Tempo

- Duration_ms

<p>&nbsp;He ignorado las variables: Liveness (si la canción esta grabada en vivo o no), Likes, Views, y Comments, ya que no describen las caracteristicas de una composición.</p>

<p>También he optado por ignorar la variable 'Key', la cual, por su naturaleza, es originalmente una variable categorica. La tonalidad representa la nota principal de una canción. A pesar de que en el conjunto de datos se represente de forma númerica, en naturaleza es una variable categorica. </p>

<p>En este caso, mi deseo es conocer cúales son las variables que mejor predicen el número de reproducciones en Spotify. Para eso voy a usar una seleccion de caracteristicas en direccion forward. Previo a esto debo hacer un <b>preprocesamiento de los datos</b>.</p>

In [6]:
variables_candidatas = spotifyYoutube.select_dtypes(include='number').drop(columns = ['Liveness', 'Key', 'Likes', 'Views', 'Comments'])
variables_candidatas.head()

Danceability  Energy  Loudness  Speechiness  Acousticness  \
0         0.818   0.705    -6.679       0.1770      0.008360   
1         0.676   0.703    -5.815       0.0302      0.086900   
2         0.695   0.923    -3.930       0.0522      0.042500   
3         0.689   0.739    -5.810       0.0260      0.000015   
4         0.663   0.694    -8.627       0.1710      0.025300   

   Instrumentalness  Valence    Tempo  Duration_ms        Stream  
0          0.002330    0.772  138.559     222640.0  1.040235e+09  
1          0.000687    0.852   92.761     200173.0  3.100837e+08  
2          0.046900    0.551  108.014     215150.0  6.306347e+07  
3          0.509000    0.578  120.423     233867.0  4.346636e+08  
4          0.000000    0.525  167.953     340920.0  6.172597e+08

### Preprocesamiento de datos

<b>· Valores nulos</b>
1. Cómo primer paso, visualizo si hay valores nulos dentro de las variables candidatas:

In [7]:
variables_candidatas.isna().sum()

Danceability          2
Energy                2
Loudness              2
Speechiness           2
Acousticness          2
Instrumentalness      2
Valence               2
Tempo                 2
Duration_ms           2
Stream              576
dtype: int64

2. Debido a que el conjunto de datos contiene <b>20718 muestras</b>, y 576 de estos valores son nulos, he decidido simplementes quitarlos, ya estos valores solo representan aproximadamente el 2.78% de los datos totales.

In [8]:
variables_candidatas = variables_candidatas.dropna()
variables_candidatas.head()

Danceability  Energy  Loudness  Speechiness  Acousticness  \
0         0.818   0.705    -6.679       0.1770      0.008360   
1         0.676   0.703    -5.815       0.0302      0.086900   
2         0.695   0.923    -3.930       0.0522      0.042500   
3         0.689   0.739    -5.810       0.0260      0.000015   
4         0.663   0.694    -8.627       0.1710      0.025300   

   Instrumentalness  Valence    Tempo  Duration_ms        Stream  
0          0.002330    0.772  138.559     222640.0  1.040235e+09  
1          0.000687    0.852   92.761     200173.0  3.100837e+08  
2          0.046900    0.551  108.014     215150.0  6.306347e+07  
3          0.509000    0.578  120.423     233867.0  4.346636e+08  
4          0.000000    0.525  167.953     340920.0  6.172597e+08

<b>· Inconsistencias en los datos</b>
 <p>&nbsp;Comprobar los valores atipicos de cada una de las variables es un proceso costoso en terminos de tiempo. Por eso primero voy a hacer la selección de caracteristicas y después comprobar valores atipicos en las variables seleccionadas y relevantes. En este paso, opte por buscar las inconsistencias más notables en los valores, en lugar de calcular valores atipicos para cada una de las variables.</p>

1. <b>Tempo</b>: En este caso una canción no puede tener un BPM de 0. Pero antes de quitarlos, quise análizar cuantos datos con un valor de 0 existen. El resultado son 17 muestras. Esto no son muchas entradas y la falta de consistencia en sus variables no le dan utilidad, por lo que (posteriormente, en una linea) voy a optar por prescindir de ellos.
<img src="https://raw.githubusercontent.com/emanuelbe1/ciencia_de_datos_repasos/main/Coderhouse/imagenes/Boxplot_Tempo.jpg"></img>

2. <p><b>Duration_ms:</b> En este caso, considere improbable que una canción tenga una duración mayor a 16 minutos. Por lo que primero quise visualizar cuales entradas representaban estos valores átipicos, y cuantas. El resultado son <b>16 muestras</b>. Si bien hay composiciones que tienen una larga duración (por ejemplo, "<i>Dogs</i>" de la banda Pink Floyd, de 17 minutos de duración) este no es el caso de las muestras en el conjunto de datos.</p>
<p>Estos valores representan "mixes", "compilados", sonidos de fondo para dormir y similares, que no son representativas de canciones. No representan un subconjunto relevante para el objetivo del análisis, por lo que he optado por eliminarlos.</p>

<img src="https://raw.githubusercontent.com/emanuelbe1/ciencia_de_datos_repasos/main/Coderhouse/imagenes/Boxplot_duration_ms_f.jpg"></img>

<p>3. &nbsp;En este paso finalmente quite los valores mencionados, y podemos ver que, después de quitar los valores nulos y los datos inconsistentes o erroneos, quedan <b>20107 muestras</b> de las 20718 originales. Este cambio en 611 entradas representaba aproximadamente un 2.95% de los datos originales. Esto no representa una perdida muy significativa.</p>

In [9]:
# Primero: extraigo las variables con un tempo mayor a 0
variables_candidatas_f = variables_candidatas[variables_candidatas['Tempo'] > 0]

print(f"Valores con un 'Tempo' mayor a cero: {variables_candidatas_f.shape[0]}")

# Luego, extraigo las variables con una duración menor a 1e6 milisegundos (esto es, 1*10⁶ = 1 millon de ms)
variables_candidatas_f = variables_candidatas_f[variables_candidatas_f['Duration_ms'] <= 1e6]
print(f"Valores con un 'Tempo' mayor a cero y menos de 16,6.. minutos de duración: {variables_candidatas_f.shape[0]}")

Valores con un 'Tempo' mayor a cero: 20123
Valores con un 'Tempo' mayor a cero y menos de 16,6.. minutos de duración: 20107


<p>&nbsp;Luego de esto, ya puedo continuar con la selección de caracteristicas o variables.</p>

### Metodo de selección de caracteristicas


<b>· Secuential Feature Selector: modelo de regresión lineal</b>
1. Teniendo el subconjunto "<i>variables_candidatas_f</i>" con potenciales variables explicativas, ahora quiero buscar cuales son las variables más adecuadas para estimar de forma lineal el número de reproducciones ("Stream") en Spotify.
 <p>Para mantener la reproductibilidad, configure el parametro <i>random_state</i> con el número 38.</p>

In [10]:
# variables_candidatas_f: son las potenciales variables explicativas, y su variable a predecir ('Stream')
# la "f" es por "filtradas": son los datos después de limpiarlos

# Defino mis caracteristicas en X y la variable objetivo en 'y'

X = variables_candidatas_f.drop(columns='Stream')
y = variables_candidatas_f[['Stream']]


# Divido mis datos en conjuntos de entrenamiento y testeo
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state = 38)

# Instancio el modelo de regresión lineal
modelo = LinearRegression()

# Uso SFS para buscar las 3 variables más explicativas segun el r²
sfs = SFS(modelo, k_features=3, forward=1, scoring='r2')

# Evaluo sobre las posibles variales
sfs.fit(X_train, y_train)

# Imprimo las tres variables con
print(f"Caracteristicas más descriptivas: {sfs.k_feature_names_}")
print(f"Valor R² después de la selección de características: {sfs.k_score_:.3f}")

Caracteristicas más descriptivas: ('Energy', 'Loudness', 'Acousticness')
Valor R² después de la selección de características: 0.026


2. El selector de caracteristicas secuencial determina que las tres variables que generan el modelo con un menor MSE o ECM son la energia, el volumen relativo y si la pieza es acustica o no. El siguiente paso es crear un modelo y medir su capacidad de predicción usando los datos de prueba.
- Al probar con cuatro caracteristicas (<i>k_features</i>) el modelo no aumento su r² de manera significativa: 0.027.
- Hay cierta mejora al elegir tres variables, las cuales generan un valor r² de 0.026.
- Al elegir dos caracteristicas el modelo lineal devuelve un r² de 0.018.

<p> Las columnas a utilizar para el modelo de regresión lineal son: 'Energy', 'Loudness', 'Acousticness'.</p>

<hr>

## <center>Modelado</center>

<center>Ajuste del modelo (regresión lineal) y configuración a partir de las caracteristicas elegidas.</center>
<p> En la sección anterior se uso un método de feature selection para reducir la dimensionalidad de los datos, en base a la regresión lineal. Las caracteristicas elegidas en base al mejor r² obtenido son:</p>

- "Energy"
- "Loudness"
- "Acousticness"

Estas columnas estan almacenadas en "variables_candidatas_f", y el objetivo es usarlas para intentar predecir el número de reproducciones.

1. Primero debo ajustar el modelo de regresión lineal al comportamiento de los datos

In [11]:
# Primero instancio el modelo de regresión lineal
modelo = LinearRegression()

# Selecciono los caracteristicas y la variable a predecir
X = variables_candidatas_f[["Energy", "Loudness", "Acousticness"]]
y = variables_candidatas_f[["Stream"]]

# Separo en entrenamiento y prueba
X_train, X_test, y_train, y_test =  train_test_split(X, y, test_size = 0.2, random_state = 38) 

# Ajusto la recta usando los datos de entrenamiento
modelo.fit(X_train, y_train)

LinearRegression()

<hr>

## <center>Validación del modelo</center>
<center> Evaluación del rendimiento del modelo e interpretación de metricas</center>
 <p>&nbsp;Debido a que el comportamiento de los datos no muestra un patrón lineal, se espera que no haya resultados de rendimiento satisfactorio. En este paso tengo que hacer una evaluación de la capacidad del modelo. Para eso voy a calcular dos metricas utiles para los modelos de regresión lineal: el MSE y el r².</p>

In [12]:
y_predichos = modelo.predict(X_test)

# Calculo el MSE
mse = mean_squared_error(y_test, y_predichos)

# Calculo el R^2
r2 = r2_score(y_test, y_predichos)

print(f'Error Cuadrado Medio (MSE): {mse:.2f}')
print(f'Coeficiente de Determinación (R^2): {r2:.2f}')

Error Cuadrado Medio (MSE): 63938717775269488.00
Coeficiente de Determinación (R^2): 0.03


<hr>

## <center>Resultados y conclusiones a partir del modelo</center>
<center> Resultados y conclusiones basados en el rendimiento del modelo </center>

<p>&nbsp;&nbsp;&nbsp;&nbsp;Según los resultados, ninguna de las variables consideradas explica de manera efectiva el número de Streams. El alto Error Cuadrático Medio (MSE) y el coeficiente de determinación R² cercano a cero indican que el modelo de regresión lineal no puede predecir la popularidad de las canciones basándose en las características seleccionadas.</p>
<p>&nbsp;Hay relaciones mucho más complejas que no son capturadas por este modelo, que quizás no pueden ser cuantificadas. Para hacer un análisis más efectivo de la posible relación que los datos presentes puedan tener con el número de reproducciones, se puede considerar aplicar modelos de agrupamiento o más avanzados que la regresión lineal, clasificación logística, o el modelo k-NN.</p>

<hr>


## <center><b>Respuestas basadas en los datos</b></center>
<br>

<b>¿Hay una relación entre el número de reproducciones en Spotfy y la adecuación para el baile de la canción?</b>

<p>&nbsp;&nbsp;&nbsp;&nbsp;La correlación entre el número de reproducciones en Spotify y la adecuación para el baile de la canción (Danceability) es de aproximadamente 0.07, lo que indica una correlación muy débil o inexistente. Esto significa que, en general, no existe una relación lineal fuerte entre estas dos variables en el conjunto de datos analizado.</p>

<p>&nbsp;Sin embargo, al hacer un análisis visual, podemos ver que al <i>atravesarse</i> los 1000 millones de reproducciones, ya no hay entran canciones con menos de 0.2 de adecuación para el baile, y los datos comienzan a presentar cierta tendencia a centrarse o dirigirse hacia 0.6 puntos de adecuación para el baile. Pasando de las 1500 millones de reproducciones, las canciones muestran un minimo de 0.3 y un maximo de 0.9 puntos de bailabilidad. Esto puede significar que se puede realizar un análisis más profundo haciendo un agrupamiento de los datos por cúalquiera de ambos ejes.</p>

<p>&nbsp;A pesar de la baja correlación, es importante señalar que otros factores podrían estar influyendo en el número de reproducciones en Spotify, como la popularidad del artista, la promoción de la canción, la fecha de lanzamiento y otros aspectos que van más allá de la adecuación para el baile. Por lo tanto, la adecuación para el baile, a primera vista, no parece ser un predictor sólido del número de reproducciones en Spotify.</p>

<br>

<b>¿Existen variables altamente correlacionadas con el número de comentarios en Youtube?</b>
<p>&nbsp;Podemos decir que las variables altamente correlacionadas al números de comentarios en un video de Youtube son los números de "Likes". Haciendo una limpieza de datos y un análisis más profundo, existe la posibilidad de que podamos considerar el número de <b>visitas</b> y <b>streams</b> cómo variables altamente correlacionadas. En resumen, diria que variables que sugieren una alta correlación seran: "<b>Views</b>", "<b>Likes</b>" y "<b>Stream</b>". Pero, en un sentido más general, la variable que se muestra (en el estado actual de los datos) más correlacionada es la variable "<b>Likes</b>".</p>

<br>

<b>¿Qué tan populares son, en términos de reproducciones en Spotify, las canciones instrumentales ("Instrumentalness' cercano a 1")?</b>

<p>&nbsp;&nbsp;&nbsp;&nbsp;Basándonos en los datos visualizados, podemos ver una tendencia interesante: las canciones que se acercan más a un valor de "1" en la métrica de "instrumentalness" tienden a tener menos reproducciones en Spotify en comparación. Esta métrica indica la probabilidad de que una canción carezca de contenido vocal. Recordando la definición de esta variable, cuanto más cercano esté "instrumentalness" a 1, mayor es la probabilidad de que la canción sea completamente instrumental y carezca de voces.</p>
<p>&nbsp;En el gráfico, podemos ver que las canciones que se acercan a un valor de "1" en "instrumentalness" generalmente tienen menos de 1 millón de reproducciones en Spotify. Sin embargo, en los datos con valores cercanos a "0" en "instrumentalness," observamos un rango más amplio de reproducciones, con algunas canciones alcanzando y superando los mil millones de reproducciones.</p>
<p>&nbsp;En resumen, nuestras visualizaciones sugieren que las canciones instrumentales tienden a tener menos reproducciones en Spotify en comparación con las canciones que contienen contenido vocal. Esto podría deberse a las preferencias del público, ya que las canciones con voces suelen ser más populares. No obstante, es importante recordar que la instrumentalidad de una canción es solo uno de los muchos factores que influyen en su popularidad en las plataformas de streaming de música.</p>

<hr>

## <center><b>Recomendaciones basadas en los descubrimientos</b></center>

<p>&nbsp;&nbsp;&nbsp;&nbsp;El análisis exploratorio de datos (EDA) realizado en este estudio ha revelado varios patrones interesantes en el conjunto de datos "<i>Youtube and Spotify</i>". A continuación, se presentan algunas recomendaciones basadas en estos hallazgos:</p>

- <b>Enfoque en la promoción de canciones instrumentales</b>: Dado que las canciones instrumentales tienden a tener menos reproducciones en Spotify en comparación con las canciones con contenido vocal, se sugiere que los artistas y las discográficas consideren estrategias de promoción específicas para este tipo de música. Esto podría incluir la identificación de audiencias nicho interesadas en música instrumental y la creación de contenido visual atractivo en plataformas como YouTube para aumentar la visibilidad.

- <b>Optimización de las características clave</b>: El análisis ha resaltado la importancia de ciertas características musicales, como la bailabilidad, la energía y el tempo. Se recomienda a artistas y compositores considerar estas características si su fin es aumentar su atractivo en plataformas de streaming y videos de YouTube.

- <b>Monitoreo de popularidad en YouTube</b>: La alta correlación entre el número de "Likes" y el número de comentarios en los videos de YouTube sugiere que la popularidad en esta plataforma puede influir en la interacción del público. Por lo tanto, se aconseja a los artistas y creadores de contenido que monitoreen de cerca sus métricas en YouTube y fomenten la interacción del público para mejorar su visibilidad.

- <b>Segmentación del público objetivo</b>: Dado que los "insights" revelados pueden variar según el género musical, se sugiere una segmentación más precisa del público objetivo. Los esfuerzos de promoción y marketing pueden adaptarse a las preferencias de diferentes audiencias, lo que aumentaría la eficacia de las estrategias de promoción.

- <b>Análisis Continuo</b>: El análisis exploratorio de datos es un proceso continuo. Se recomienda que se sigan analizando los datos a medida que se recopilan más observaciones. Esto permitirá mantenerse al día con las tendencias cambiantes y ajustar las estrategias en consecuencia.

<p>&nbsp;Estas recomendaciones se basan en los "insights" descubiertos durante el EDA y pueden ayudar a los profesionales de la industria musical, artistas y creadores de contenido a tomar decisiones más informadas para optimizar su presencia en plataformas como Spotify y YouTube.</p>